src/server/pfs/server/source.go

```diff
@@ -0,0 +1,255 @@
+	"github.com/pachyderm/pachyderm/src/client/pfs"
+	"github.com/pachyderm/pachyderm/src/client/pkg/errors"
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset"
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset/index"
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset/tar"
+	"golang.org/x/net/context"
+)
+
+// FileReader is a PFS wrapper for a fileset.MergeReader.
+// The primary purpose of this abstraction is to convert from index.Index to
+// pfs.FileInfo and to convert a set of index hashes to a file hash.
+type FileReader struct {
+	file      *pfs.File
+	idx       *index.Index
+	fmr       *fileset.FileMergeReader
+	mr        *fileset.MergeReader
+	fileCount int
+	hash      hash.Hash
+}
+
+func newFileReader(file *pfs.File, idx *index.Index, fmr *fileset.FileMergeReader, mr *fileset.MergeReader) *FileReader {
+	h := pfs.NewHash()
+	for _, dataRef := range idx.DataOp.DataRefs {
+		// TODO Pull from chunk hash.
+		h.Write([]byte(dataRef.Hash))
+	}
+	return &FileReader{
+		file: file,
+		idx:  idx,
+		fmr:  fmr,
+		mr:   mr,
+		hash: h,
+	}
+}
+
+func (fr *FileReader) updateFileInfo(idx *index.Index) {
+	fr.fileCount++
+	for _, dataRef := range idx.DataOp.DataRefs {
+		fr.hash.Write([]byte(dataRef.Hash))
+	}
+}
+
+// Info returns the info for the file.
+func (fr *FileReader) Info() *pfs.FileInfoV2 {
+	return &pfs.FileInfoV2{
+		File: fr.file,
+		Hash: pfs.EncodeHash(fr.hash.Sum(nil)),
+	}
+}
+
+// Get writes a tar stream that contains the file.
+func (fr *FileReader) Get(w io.Writer, noPadding ...bool) error {
+	if err := fr.fmr.Get(w); err != nil {
+		return err
+	}
+	for fr.fileCount > 0 {
+		fmr, err := fr.mr.Next()
+		if err != nil {
+			return err
+		}
+		if err := fmr.Get(w); err != nil {
+			return err
+		}
+		fr.fileCount--
+	}
+	if len(noPadding) > 0 && noPadding[0] {
+		return nil
+	}
+	// Close a tar writer to create tar EOF padding.
+	return tar.NewWriter(w).Close()
+}
+
+func (fr *FileReader) drain() error {
+	for fr.fileCount > 0 {
+		if _, err := fr.mr.Next(); err != nil {
+			return err
+		}
+		fr.fileCount--
+	}
+	return nil
+}
+
+// Source iterates over FileInfoV2s generated from a fileset.ReaderAPI
+type Source struct {
+	commit        *pfs.Commit
+	getReader     func() fileset.FileSource
+	computeHashes bool
+}
+
+// Creates a Reader which emits FileInfoV2s with the information from commit, and the entries from readers
+// returned by getReader.  If getReader returns different Readers all bets are off.
+func NewSource(commit *pfs.Commit, computeHashes bool, getReader func() fileset.FileSource) *Source {
+	return &Source{
+		commit:        commit,
+		getReader:     getReader,
+		computeHashes: computeHashes,
+	}
+}
+
+func (s *Source) Iterate(ctx context.Context, cb func(*pfs.FileInfoV2, fileset.File) error) error {
+	ctx, cf := context.WithCancel(ctx)
+	defer cf()
+	fs1 := s.getReader()
+	fs2 := s.getReader()
+	s2 := newStream(ctx, fs2)
+	cache := make(map[string][]byte)
```

Comment:
Will change this to `FileInfoV2`s with updates to the protobuf for size and whatever else.  I made a note with that item in the doc.